### PR TITLE
WIP: todo.txt: New date picker 

### PR DESCRIPTION
**Pull request by harshad1**
In this PR I have updated the todo txt date action to use a date picker.

1. Short press opens a date picker with the current date
2. Long press will default to 'due:<today's date> + 3'
3. If a date is selected, the date will be parsed and used as the selected date. This is useful as  android phones will highlight a date on long-press. So to change a date you can long-press date to highlight it, and then press date button to change it.

The datepicker uses a FragmentManager to manage the dialog as per Android dev recommendations.

__Todo:__
- [x] Determine if we should change the long press action to insert today (instead of today + 3)
- [x] Determine if we can / should modify the 'current' date if the cursor is on a date when the dialog is launched

### CODE REVIEW

**Comment by harshad1** [comment]
IMO the `attach->date` in markdown should default to a simple date picker and an 'advanced' button should launch the `DatetimeFormatDialog`

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477&bid=18279392)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java [hunk 1]
```diff
@@ -10,16 +10,22 @@
...
 import android.support.annotation.DrawableRes;
 import android.support.annotation.StringRes;
+import android.support.v4.app.DialogFragment;
+import android.support.v4.app.FragmentActivity;
 import android.support.v4.content.ContextCompat;
 import android.support.v7.widget.TooltipCompat;
 import android.text.Editable;
 import android.view.View;
 import android.view.ViewGroup;
+import android.widget.DatePicker;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java:28]: Codacy Issue found: [Unused import - android.widget.DatePicker.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477)

Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java:28]: Codacy Issue found: [Avoid unused imports such as 'android.widget.DatePicker'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477)

**Comment by gsantner** [comment]
Hey thanks @harshad1 ..but we already have a date/time dialog? (The one from "attachments" menu)

I suggest to rather keep one implementation. What do you think?

**Comment by harshad1** [comment]
> Hey thanks @harshad1 ..but we already have a date/time dialog? (The one from "attachments" menu)
> 
> I suggest to rather keep one implementation. What do you think?

What I've added only does date picking while the existing one does date and time and more advanced formatting, using the current dialog is not very efficient to insert a date in todotxt. I see your point about duplicated functionality though.

Let me see if I can combine the functionality in a clean way.

File: app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java [hunk 2]
```diff
@@ -31,18 +37,26 @@
...
 import java.util.Date;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
 import java.util.Set;
+import java.util.regex.Matcher;
+
+import butterknife.OnClick;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java:57]: Codacy Issue found: [Unused import - butterknife.OnClick.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477&bid=18279392)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java (hunk 2 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java:57]: Codacy Issue found: [Avoid unused imports such as 'butterknife.OnClick'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477)

File: app/src/main/java/net/gsantner/markor/format/general/DatetimeFormatDialog.java [hunk 3]
```diff
@@ -94,8 +94,6 @@ public static void showDatetimeFormatDialog(final Activity activity, final Highl
         final Button datePickButton = viewRoot.findViewById(R.id.start_datepicker_button);
         final Button timePickButton = viewRoot.findViewById(R.id.start_timepicker_button);
         final CheckBox formatInsteadCheckbox = viewRoot.findViewById(R.id.get_format_instead_date_or_time_checkbox);
-        final CheckBox alwaysNowCheckBox = viewRoot.findViewById(R.id.always_use_current_datetime_checkbox);
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/format/general/DatetimeFormatDialog.java:97]: Removed `alwaysNowCheckBox`. This isn't really useful. A 'set now' button might be better.

File: app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java [hunk 4]
```diff
@@ -536,4 +548,146 @@ public ActionItem(int[] data) {
...
+            }
+
+            // Add requested offset
+            calendar.add(Calendar.DATE, deltaDays);
+
+            DateTimeFragment dateFragment = new DateTimeFragment()
+                    .setActivity(_activity)
+                    .setDateListener(listener)
+                    .setCalendar(calendar)
+                    .setExtra("Advanced", new DialogInterface.OnClickListener() {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java:591]: Added an 'advanced' button which kicks you onto the advanced date and time dialog

**Comment by harshad1** [comment]
@gsantner After thinking about this some more, here is what I feel:
1. We need a fast (2 clicks at most) way to add dates in todo.txt
2. Todo.txt only really supports YYYY-MM-DD. Changing formats is not essential.
2. Todo.txt requires fast date changing and parsing of the tag:YYYY-MM-DD format
2. Markdown requires more complex dates and time support  

For now, I think we should use the dialog I have added for todo.txt alone while using the existing DateTimeFormatDialog for markdown / plaintext etc. We should work on streamlining the DateTimeFormatDialog separately.

File: app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java [hunk 5]
```diff
@@ -10,11 +10,16 @@
 package net.gsantner.markor.ui.hleditor;
 
 import android.app.Activity;
+import android.app.DatePickerDialog;
+import android.app.Dialog;
 import android.content.Context;
+import android.content.DialogInterface;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java:16]: Codacy Issue found: [Avoid unused imports such as 'android.content.DialogInterface'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477&bid=18279392)

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java [hunk 6]
```diff
@@ -10,14 +10,18 @@
...
 import android.support.annotation.StringRes;
+import android.support.v4.app.FragmentActivity;
 import android.text.Editable;
 import android.text.TextUtils;
 import android.view.HapticFeedbackConstants;
 import android.view.View;
 
 import net.gsantner.markor.R;
 import net.gsantner.markor.format.general.CommonTextActions;
+import net.gsantner.markor.format.general.DatetimeFormatDialog;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:24]: Codacy Issue found: [Unused import - net.gsantner.markor.format.general.DatetimeFormatDialog.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477)

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java [hunk 7]
```diff
@@ -10,14 +10,18 @@
 package net.gsantner.markor.format.todotxt;
 
 import android.app.Activity;
+import android.app.DatePickerDialog;
+import android.content.DialogInterface;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:14]: Codacy Issue found: [Unused import - android.content.DialogInterface.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477)

Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:14]: Codacy Issue found: [Avoid unused imports such as 'android.content.DialogInterface'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477)

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java (hunk 6 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:24]: Codacy Issue found: [Avoid unused imports such as 'net.gsantner.markor.format.general.DatetimeFormatDialog'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java (hunk 5 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java:16]: Codacy Issue found: [Unused import - android.content.DialogInterface.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477)

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java [hunk 8]
```diff
@@ -322,4 +330,71 @@ public boolean onLongClick(View v) {
...
+     * the todo.txt date format `tag:YYYY-mm-dd` or plain `YYYY-mm-dd`.
+     * If a date is found, it is parsed and displayed to the user to be updated.
+     *
+     * If no date is found, the user selected date will be inserted at the cursor.
+     * If provided, a key will be inserted before the date.
+     * (key = 'due => date inserted = `due:YYYY-mm-dd`)
+     *
+     * @param key An optional prefix key. optional (use null for no key)
+     */
+    protected void updateOrInsertTodoDate(final String key) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:347]: Suggestion for simplification:
* Use `String oldDue = SttCommander::parseDueDate()` function on task of current line
* String currentLine = ....; currentLine = currentLine.replace(oldDue, newDue);

**Comment by gsantner** [comment]
yes please just keep going, have the new date stuff specific to todo.txt yyyy-mm-dd dates.

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java (hunk 8 above)
Comment (harshad1) [reply app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java]: The function searches for a date _under the cursor_.

My use cases are
1. Change creation date (rare)
1. Change due date
2. Change threshold date

With the current approach, I can tap on a date, tap the 'date' action button and quickly change the date.

I think what you are suggesting is that we detect and insert or replace the due date in the current line. Am I correct?

**Comment by harshad1** [comment]
There are now two functions. 
- `updateOrInsertSelectedDate`: inserts a plain date. If the currently highlighted text is a date, the date is parsed and used as the initial date. This can be used to quickly change dates
    - Long press date to select it
    - Tap date action
    - Pick new date and click ok to change
- `updateOrAppendDueDate`: Searches the current line for due:YYYY-mm-dd and presents it to be updated. If not found, it is appended to the line.

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java [hunk 9]
```diff
@@ -29,11 +34,18 @@
...
+import net.gsantner.opoc.util.StringUtils;
 
 import java.io.File;
+import java.text.ParseException;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Calendar;
 import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:46]: Codacy Issue found: [Unused import - java.util.regex.Pattern.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477&bid=18279392)

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java (hunk 9 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:46]: Codacy Issue found: [Avoid unused imports such as 'java.util.regex.Pattern'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5575477)

File: app/src/main/java/net/gsantner/opoc/format/todotxt/SttCommander.java [hunk 10]
```diff
@@ -62,6 +62,7 @@
     public static final Pattern PATTERN_PRIORITY_F = Pattern.compile("(?:^|\\n)\\(([Ff])\\)\\s");
     public static final Pattern PATTERN_COMPLETION_DATE = Pattern.compile("(?:^|\\n)(?:[Xx] )(" + PT_DATE + ")");
     public static final Pattern PATTERN_CREATION_DATE = Pattern.compile("(?:^|\\n)(?:\\([A-Za-z]\\)\\s)?(?:[Xx] " + PT_DATE + " )?(" + PT_DATE + ")");
+    public static final Pattern PATTERN_IS_DATE = Pattern.compile(PT_DATE);
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/opoc/format/todotxt/SttCommander.java:65]: The pattern catches a date, `_IS_` is something that is done by using the respective methods. Suggest to rename to just PATTERN_DATE

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java [hunk 11]
```diff
@@ -322,4 +334,151 @@ public boolean onLongClick(View v) {
...
+            super();
+            setCalendar(Calendar.getInstance());
+        }
+
+        public DateFragment setListener(DatePickerDialog.OnDateSetListener listener) {
+            this.listener = listener;
+            return this;
+        }
+
+        public DateFragment setActivity(Activity activity) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:452]: isn't this info exposed via `getActivity()` already as long attached to a activity?

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java [hunk 12]
```diff
@@ -322,4 +334,151 @@ public boolean onLongClick(View v) {
...
+        private int month;
+        private int day;
+
+        public DateFragment() {
+            super();
+            setCalendar(Calendar.getInstance());
+        }
+
+        public DateFragment setListener(DatePickerDialog.OnDateSetListener listener) {
+            this.listener = listener;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:448]: please use `_member` for members, shorter than `this.` every time

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java [hunk 13]
```diff
@@ -322,4 +334,151 @@ public boolean onLongClick(View v) {
...
+            else calendar.add(Calendar.DAY_OF_MONTH, offset);
+        } catch (ParseException e) {
+            // Regex failed?; should not be here
+            e.printStackTrace();
+        }
+
+        DatePickerDialog.OnDateSetListener listener = (view, year, month, day) -> {
+            Calendar fmtCal = Calendar.getInstance();
+            fmtCal.set(year, month, day);
+            String tagDate = String.format("due:%s", DATEF_YYYY_MM_DD.format(fmtCal.getTime()));
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:412]: lint will tells with locale here right? Then just add root locale

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java [hunk 14]
```diff
@@ -322,4 +334,151 @@ public boolean onLongClick(View v) {
...
+
+        Matcher match = SttCommander.PATTERN_DUE_DATE.matcher(line);
+        final boolean found = match.find();
+
+        Calendar calendar = Calendar.getInstance();
+        // Parse selection for date
+        try {
+            if (found) calendar.setTime(DATEF_YYYY_MM_DD.parse(match.group(0)));
+            else calendar.add(Calendar.DAY_OF_MONTH, offset);
+        } catch (ParseException e) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:404]: please make it 

```
if () {

} else {

}
```

File: app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java [hunk 15]
```diff
@@ -322,4 +334,151 @@ public boolean onLongClick(View v) {
...
+        final int[] selection = StringUtils.getSelection(_hlEditor);
+        final Editable text = _hlEditor.getText();
+        final CharSequence dateText = text.subSequence(selection[0], selection[1]);
+
+        Calendar calendar = Calendar.getInstance();
+
+        // Parse selection for date
+        try {
+            final Matcher match = SttCommander.PATTERN_IS_DATE.matcher(dateText);
+            if (match.find()) calendar.setTime(DATEF_YYYY_MM_DD.parse(match.group(0)));
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java:357]: if ()  {

**Comment by harshad1** [comment]
@harshad1 Thanks for the feedback. I am taking this pr down for now. I will probably make another one with simpler due/threshold date setting using SttTask.

**Comment by gsantner** [comment]
You can keep it open

**Pull request closed without merging**